larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,892 @@
+    const CartesianVector rhsDistanceVector(pRhs->GetPositionVector() - m_startPoint);
+
+    const float lhsProjectedDistance(lhsDistanceVector.GetDotProduct(m_lineDirection));
+    const float rhsProjectedDistance(rhsDistanceVector.GetDotProduct(m_lineDirection));
+
+    return (lhsProjectedDistance < rhsProjectedDistance);
+}
+ 
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+TrackInEMShowerAlgorithm::TrackInEMShowerAlgorithm() :
+    m_minCaloHits(25), 
+    m_minSeparationDistance(27.f),
+    m_maxPredictedMergePointOffset(5.f),
+    m_slidingFitWindow(20),
+    m_mergePointMinCosAngleDeviation(0.9995),
+    m_minClusterLengthSum(75.f),
+    m_minDirectionDeviationCosAngle(0.99),
+    m_distanceFromLine(0.35),
+    m_maxTrackGaps(2),
+    m_lineSegmentLength(7.f),
+    m_maxHitDistanceFromCluster(4.f),
+    m_maxHitSeparationForConnectedCluster(4.f),
+    m_maxDistanceFromMainTrack(0.75)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode TrackInEMShowerAlgorithm::Run()
+{
+    const ClusterList *pClusterList = NULL;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pClusterList));
+
+    const CaloHitList *pCaloHitList = NULL;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pCaloHitList));
+    
+    ClusterVector clusterVector;
+    this->SelectCleanClusters(pClusterList, clusterVector);
+
+    TwoDSlidingFitResultMap microSlidingFitResultMap;
+    TwoDSlidingFitResultMap macroSlidingFitResultMap;
+    this->InitialiseSlidingFitResultMaps(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    // Find longest length association and conitnue merging until no further merges can be made
+    bool mergeMade(false);
+    do
+    {
+        ClusterAssociation clusterAssociation;
+        if(!this->FindBestClusterAssociation(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap, clusterAssociation))
+            break;
+        
+        ClusterToCaloHitListMap clusterToCaloHitListMap;
+        CaloHitVector extrapolatedCaloHitVector;
+        this->GetExtrapolatedCaloHits(clusterAssociation, pClusterList, extrapolatedCaloHitVector, clusterToCaloHitListMap);
+        
+        if (extrapolatedCaloHitVector.empty())
+            break;
+        
+        if (!this->IsTrackContinuous(clusterAssociation, extrapolatedCaloHitVector))
+            break;
+        
+        this->RefineTracks(clusterAssociation, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector, extrapolatedCaloHitVector);
```

Comment:
Is it possible to call MergeHits first (which internally does all required updates for deletion, etc.), then RefineTrack for the whole track in one go, or does that break something? (RefineTrack could internally call the functions to do required updates for deletion, etc. again.)